### PR TITLE
[#9] Helper for filter form

### PR DESCRIPTION
**Pull request by maltoe**
@woylie Here's yesterday's snippet in `Flop.Phoenix` components.. No tests yet, basically a proof of concept for further discussion. Main point I'm unsure about is still the `inputs_for` vs. lower-level decision making (see comment).

please take a 👀

### CODE REVIEW

File: lib/flop_phoenix.ex [hunk 1]
```diff
@@ -467,6 +466,101 @@ defmodule Flop.Phoenix do
...
+
+  ## Assigns
+
+  - `form` - `t:Phoenix.HTML.FormData`
+  - `label` (optional) - label to display, defaults to field name
+  - `using` - helper to render primary input field
+  - `op_selectable` (optional) - defines whether operation can be selected from
+    list, defaults to `false`
+  - `op_selectable_from` (optional) - list of `t:Flop.Filter.op` operations to
+    choose from
```
[the remark this answers is not in the document]
Comment (maltoe) [reply lib/flop_phoenix.ex:480]: That's an excellent idea!

File: lib/flop_phoenix.ex [hunk 2]
```diff
@@ -467,6 +466,101 @@ defmodule Flop.Phoenix do
...
+  """
+  @spec filter_input(map) :: Phoenix.LiveView.Rendered.t()
+  def filter_input(assigns) do
+    assigns = Filter.init_input_assigns(assigns)
+
+    ~H"""
+    <div>
+      <%= hidden_input @form, :field %>
```
[the remark this answers is not in the document]
Comment (maltoe) [reply lib/flop_phoenix.ex:488]: re: `filters[]` -> I think this is why I'm still confused about the `filter_input` vs `filter_inputs_for` approach, in particular how these two public APIs would combine in practise.

If you have a form like this:

```
<input name="filters[][field]" value="foo">
<input name="filters[][op]" value="==">
<input name="filters[][value]" value="bar">
<input name="filters[][field]" value="baz">
<input name="filters[][op]" value="==">
<input name="filters[][value]" value="quuz">
```

the params Phoenix makes of this are 

```
%{"filters" => [%{"field" => "foo"}, %{"op" => "=="}, %{"value" => "bar"}, %{"field" => "baz"}, %{"op" => "=="}, %{"value" => "quuz"}]}
```

which is not what we want. We want a form like this:

```
<input name="filters[0][field]" value="foo">
<input name="filters[0][op]" value="==">
<input name="filters[0][value]" value="bar">
<input name="filters[1][field]" value="baz">
<input name="filters[1][op]" value="==">
<input name="filters[1][value]" value="quuz">
```

... which of course we can also achieve for `filter_input/1` by giving it a manual `index` option, so

```
<.filter_input form={f} field="foo" index="0" meta={meta} />
```

renders as the first the lines above (by looking up the `"bar"` value in the `meta`). But I doubt this is what you had in mind? We can totally follow through with this approach, but TBH then the outer form doesn't need to use `inputs_for` anymore as we entirely bypass its internal enumeration+naming logic, as well as the data fetching. 

In other words, the more I think about it, the less is the `filter_input` API you have in mind a part of `filter_inputs_for`, it basically doesn't have use for it when it uses `Phoenix.HTML.Form.inputs_for`.

---

May I suggest the following: We continue with `filter_inputs_for` and *don't* make `filter_input` a public API for now, to focus on one of the envisioned ways to create these forms. And we revisit a low-level public `filter_input` once we have all the associated (shared) boilerplate in place? I think this would simplify the first iteration of this quite a bit.

File: lib/flop_phoenix/filter.ex [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+  end
+
+  defp default_input(type) when type in [:id, :binary_id, :string, :binary],
+    do: :text_input
+
+  defp default_input(type) when type in [:integer, :float, :decimal],
+    do: :number_input
+
+  defp default_input(:boolean), do: :checkbox
+  defp default_input(:date), do: :date_input
```
[the remark this answers is not in the document]
Comment (maltoe) [reply lib/flop_phoenix/filter.ex:58]: of course, sorry, not sure how this ended up here

File: lib/flop_phoenix/filter.ex [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  defp assign_op_selectable_from(%{op_selectable_from: _list} = assigns),
+    do: assigns
+
+  defp assign_op_selectable_from(assigns) do
+    assign(assigns, :op_selectable_from, default_op_selectable_from())
+  end
+
+  defp default_op_selectable_from do
+    # TODO: For a sensible preselection we need the type again here.
+    # But I tried to keep `filter_input` agnostic of the type (no `for` assign).
```
[the remark this answers is not in the document]
Comment (maltoe) [reply lib/flop_phoenix/filter.ex:158]: See the other comment above, that perhaps makes it clear. Basically I wanted to keep the "heavy lifting" in our component (= setting defaults, looking up field type & determining default input type from it, getting data from current meta) and let the other one relatively dumb. In the end I think the `inputs_for` approach doesn't align well with a public API `filter_input`

**Comment by maltoe** [comment]
@woylie Thanks for your review! As outlined in the comment above, I'm still struggling to fully grasp your vision of the `filter_input` API and the `filter_inputs_for` API (with a list of fields). I simplified the current code a bit and dropped the public `filter_input` for now. Please let me know what you think, if I should proceed (adding features from your list, writing tests), or if you want to have a go on it.

**Comment by maltoe** [comment]
I'm also happy to have a call on this topic, if you are up for it. I'm in CEST, if you hadn't guessed already.

**Comment by woylie** [comment]
Hey @maltoe, thanks for the update. I'll need to do some exploration before I can come back to you.

**Comment by maltoe** [comment]
@woylie Sure, take your time 👍  

One last thing, which came to me yesterday after working on this:

> And since you pointed out that filter_inputs_for could be called multiple times with different sets of filters (and possibly different builder functions), I wonder whether this lowest level API is still needed in the end.

This is of course not true, as unfortunately multiple calls to `filter_inputs_for` (in its current version) would each start over indexing the nested maps at `0`. For instance

```
<.filter_inputs_for fields={[:foo]} ... />
<.filter_inputs_for fields={[:bar]} ... />
```

generates this HTML:

```
...
<input name="filters[0][field]" value="foo">
...
<input name="filters[0][field]" value="bar">
```

so in the resulting parameters `foo` will be replace by `bar`. So this is actually not an option for "specializing" certain inputs.

---

Also feel free to hijack this PR if it helps!
